tlatools/org.lamport.tlatools/src/tlc2/value/impl/FcnRcdValue.java

```diff
@@ -381,76 +379,80 @@ public final Value select(Value arg) {
-          for (int i = 0; i < len; i++) {
-            if (this.domain[i].equals(arg)) {
-              return this.values[i];
-            }
-          }
-        }
+    	  return selectBinarySearch(arg);
       }
-      return null;
-
     }
     catch (RuntimeException | OutOfMemoryError e) {
       if (hasSource()) { throw FingerprintException.getNewHead(this, e); }
       else { throw e; }
     }
   }
+  
+  // This implementation has a concurrency bug: https://github.com/tlaplus/tlaplus/issues/439
+//  final Value selectIndexTable(final Value arg) {
+//    // domain is represented as an array of values:
+//    if (this.indexTbl != null) {
+//      return this.values[this.lookupIndex(arg)];
+//    }
+//    if (this.isNorm) {
+//    	this.createIndex();
+//    }
+//    if (this.indexTbl != null) {
+//      return this.values[this.lookupIndex(arg)];
+//    }
+//    else {
+//	  return selectLinearSearch(arg);
+//    }
+//  }
 
-  public final Value selectNoIndex(Value arg) {
-    try {
-
-      if (this.intv != null) {
-        // domain is represented as an integer interval:
-        if (!(arg instanceof IntValue)) {
-          Assert.fail("Attempted to apply function with integer domain to" +
-                " the non-integer argument " + Values.ppr(arg.toString()), getSource());
-        }
-        int idx = ((IntValue)arg).val;
-        if ((idx >= this.intv.low) && (idx <= this.intv.high)) {
-          return this.values[idx - this.intv.low];
+  final Value selectLinearSearch(final Value arg) {
+      // domain is represented as an array of values:
+      int len = this.domain.length;
+      for (int i = 0; i < len; i++) {
+        if (this.domain[i].equals(arg)) {
+          return this.values[i];
         }
       }
-      else {
-        // domain is represented as an array of values:
-          int len = this.domain.length;
-          for (int i = 0; i < len; i++) {
-            if (this.domain[i].equals(arg)) {
-              return this.values[i];
-            }
-          }
-      }
       return null;
-    }
-    catch (RuntimeException | OutOfMemoryError e) {
-      if (hasSource()) { throw FingerprintException.getNewHead(this, e); }
-      else { throw e; }
-    }
   }
 
+  final Value selectBinarySearch(Value arg) {
+	  // The value 32 has been determined empirically (see FcnRcdBenachmark).
+	  // In older versions of TLC this the threshold was 10.
+      if (this.isNorm && this.domain.length >= LINEAR_SEARCH_THRESHOLD) {
+        // domain is represented as an array of values:
+    	// iff normalized, use binary search to speedup lookups.
+		int low = 0;
+		int high = this.domain.length - 1;
+		while (low <= high) {
+			// cast to int is the same as flooring for positive ints.
+			final int mid = low + (high - low) / 2;
+			final int cmp = this.domain[mid].compareTo(arg);
+			if (cmp == 0) {
+				return this.values[mid];
+			} else if (cmp < 0) {
+				low = mid + 1;
+			} else {
+				high = mid;
+			}
+		}
+		return null;
```

Comment:
[`java.util.Arrays.binarySearch()`](https://docs.oracle.com/javase/8/docs/api/java/util/Arrays.html#binarySearch-java.lang.Object:A-java.lang.Object-) already implements binary search over arrays. Is there a good reason to use a homebrewed implementation here?